modules/storages/app/services/storages/file_link_sync_service.rb

```diff
@@ -0,0 +1,218 @@
+  #    "mtime" : 1655301278,      # Linux epoch file modification +overwrite
+  #    "mimetype" : "application/pdf",  # +overwrite
+  #    "name" : "Nextcloud Manual.pdf", # "Canonical" name, could changed by owner +overwrite
+  #    "owner_id" : "admin",      # ID at Nextcloud side +overwrite
+  #    "owner_name" : "admin",    # Name at Nextcloud side +overwrite
+  #    "size" : 12706214,         # Not used yet in OpenProject +overwrite
+  #    "status" : "OK",           # Not used yet
+  #    "statuscode" : 200,        # Not used yet
+  #    "trashed" : false          # Exclude trashed files from result array of FileLinks
+  # }
+  # In case of permission errors (depending on the current user) we get:
+  # "24" => { "status" => "Forbidden", "statuscode" => 403 }
+  # In case of completely deleted file or other errors we might also get:
+  # "24" = { "status" => "Not Found", "statuscode" => 404 }
+  def sync_single_file(file_link, origin_file_info_hash)
+    file_link.origin_mime_type = origin_file_info_hash["mimetype"]
+    file_link.origin_created_by_name = origin_file_info_hash["owner_name"]
+    file_link.origin_name = origin_file_info_hash["name"]
+    file_link.origin_created_at = Time.zone.at(origin_file_info_hash["ctime"])
+    file_link.origin_updated_at = Time.zone.at(origin_file_info_hash["mtime"])
+
+    file_link
+  end
+
+  private
+
+  # Sync a Nextcloud storage
+  def sync_nextcloud(storage_id, storage_file_links)
+    # Get the OAuthClientToken that will authenticate us against Nextcloud
+    oauth_client = OAuthClient.find_by(integration_id: storage_id)
+    connection_manager = ::OAuthClients::ConnectionManager.new(user: @user, oauth_client:)
+    connection_manager_service_result = connection_manager.get_access_token # No scope for Nextcloud...
+    unless connection_manager_service_result.success
+      set_error_for_file_links(storage_file_links)
+      return
+    end
+
+    @oauth_client_token = connection_manager_service_result.result
+
+    storage_origin_file_ids = storage_file_links.map(&:origin_id)
+    nextcloud_request_result = connection_manager.request_with_token_refresh do
+      response = request_files_info(storage_origin_file_ids)
+      # Parse HTTP response an return a ServiceResult with:
+      #   success: result= Hash with Nextcloud filesinfo (name of endpoint) data
+      #   failure: result= :error or :not_authorized
+      parse_files_info_response(response)
+    end
+    # Pass errors from Nextcloud into service result
+    @service_result.merge!(nextcloud_request_result)
+
+    if nextcloud_request_result.failure?
+      set_error_for_file_links(storage_file_links)
+      return
+    end
+
+    set_file_link_permissions(storage_file_links, nextcloud_request_result.result)
+  end
+
+  def set_error_for_file_links(storage_file_links)
+    @service_result.result += storage_file_links.each { |file_link| file_link.origin_permission = :error }
+    @service_result.success = false
```

Comment:
not sure, if I like that concept, that the service result shows a failure, just because there is no access token for a single storage. I mean, the user won't see it, but another developer can easily get confused by this fact.

Reply:
not to be solved here ...